### PR TITLE
Cryptsetup fixes

### PR DESCRIPTION
**Pull request by fbuihuu**

### CODE REVIEW

**Comment by fbuihuu** [comment]
@poettering I still find "extrinsic" pretty obscure from a user POV but if you still insist I can rename the option.

File: src/cryptsetup/cryptsetup-generator.c [hunk 1]
```diff
@@ -238,6 +238,7 @@ static int create_disk(
         tmp = fstab_test_option(options, "tmp\0");
         swap = fstab_test_option(options, "swap\0");
         netdev = fstab_test_option(options, "_netdev\0");
+        create_in_initrd = fstab_test_option(options, "x-initrd.create\0");
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup-generator.c]: This would work, but I think we should be able to make that automatic. Anything that is a dependency of `-.mount` should be excluded from killing. This makes yet another bit of internal state visible and configurable from the outside.

What about the following: on reload, create a drop-in that cancels the Conflicts=shutdown.target setting for the device configured as backing for `-.mount`? I'm not sure about the best implementation, but maybe create
the conflicts in a special `dev-<foo>.d/10-default-shutdown.conf` from cryptsetup target, and override that with `dev-<foo>.d/20-default-shutdown.conf` from any other target that configures `-.mount` or `usr.mount`, etc.

File: src/cryptsetup/cryptsetup-generator.c [hunk 2]
```diff
@@ -290,7 +290,7 @@ static int create_disk(
                 "Documentation=man:crypttab(5) man:systemd-cryptsetup-generator(8) man:systemd-cryptsetup@.service(8)\n"
                 "SourcePath=%s\n"
                 "DefaultDependencies=no\n"
-                "Conflicts=umount.target\n"
+                "Conflicts=shutdown.target\n"
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-generator.c]: not sure i grok this change. I mean, doing "systemctl start umount.target" means all mounts are taken down, and doing "systemctl start shutdown.target" means all services are taken down. While taking down all mounts kinda implies taking down most services the opposite is not true: i.e. taking down all services shouldn't mean that all mounts are taken down too. Now, with your change you suddenly make it so that taking down all services also means detaching the storage behind mounts without necessarily taking down the mounts, and that doesn't really make sense, does it?

I mean, I think in the end these two targets are nor as independent as one might like, so the discussion is moot, but it's not clear to me why to have this commit hence. Can you elaborate?

File: src/cryptsetup/cryptsetup-generator.c [hunk 3]
```diff
@@ -227,7 +227,7 @@ static int create_disk(
                 *filtered = NULL, *u_escaped = NULL, *filtered_escaped = NULL, *name_escaped = NULL, *header_path = NULL;
         _cleanup_fclose_ FILE *f = NULL;
         const char *dmname;
-        bool noauto, nofail, tmp, swap, netdev;
+        bool noauto, nofail, tmp, swap, netdev, create_in_initrd;
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-generator.c]: i think adding this mount option mirroring x-initrd.mount makes a lot of sense. But I'd really cal this "x-initrd.attach" or so, after all we usually talk about attaching LUKS volumes, not "creating" them. (i.e. creating a LUKS volume probably would mostly understand as "formatting something as LUKS")

File: src/cryptsetup/cryptsetup-generator.c (hunk 1 above)
Comment (poettering) [reply src/cryptsetup/cryptsetup-generator.c]: the problem with these automatic deps is that DM devices have multiple names and /proc/self/mountinfo is likely going to show a generic one (i.e. /dev/md-0 rather than /dev/mapper/luks-b5d403b9-344c-43e9-9fbc-daa8596f204c), hence we are likely to not match things up correctly. I fear tagging something manually is something we cannot avoid... (or to turn this arround: maybe add the automatic mechanism too, but the explicit one like @fbuihuu implemented it in this PR is something to have as well, already because people might do crazy stuff like splitting of /usr/lib/share and thus might need to mark more stuff as "mount in initrd")

File: units/system-systemd\x2dcryptsetup.slice [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+#
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+[Slice]
+DefaultDependencies=no
```
Comment (poettering) [inline units/system-systemd\x2dcryptsetup.slice:13]: while you ar at it, please set a nice Description=

File: src/cryptsetup/cryptsetup-generator.c (hunk 1 above)
Comment (fbuihuu) [reply src/cryptsetup/cryptsetup-generator.c]: @keszybz I agree with @poettering: even if we could make it automatic for / and /usr we still need the option to deal with other paths that need to be mounted from initrd (and yes we already have some use cases here with transactional systems).

That said I'm not sure to understand your proposal. The drop-ins adding conflicts would be added during reload, which makes the whole thing depends on the fact that a reload is done before rebooting. It's very likely indeed but still seems a bit fragile no ?

Also shouldn't the conflicts have to be added to the systemd-cryptsetup services ?

File: src/cryptsetup/cryptsetup-generator.c (hunk 2 above)
Comment (fbuihuu) [reply src/cryptsetup/cryptsetup-generator.c]: I can see your point.

When I opted for this change I considered the following points:

- systemd-cryptsetup@*.service are services so they should be stopped by shutdown.target in theory. But as you said it doesn't really make sense since stopping all services doesn't necessarily mean unmounting all filesystems.

- IMU, doing "systemctl start umount.target" is supposed to deal with mount units only, it isn't supposed to stop systemd-cryptsetup@.service and detach the disks behind mounts. Unmounting doesn't mean that the disk behind the mount should be removed, ejected or detached usually, does it ?

I'll drop this patch, however it still seems to me that conflicting with umount.target isn't correct.

Maybe the cryptsetup.target should have its counterpart target for shutting down crypto devices like umount.target is for mount.target.

**Comment by fbuihuu** [comment]
@poettering @keszybz so I dropped the patch that made the cryptsetup services conflict with shutdown.target and addressed the different points made by @poettering.

PTAL

**Comment by poettering** [comment]
lgtm. @keszybz ?

File: man/crypttab.xml [hunk 5]
```diff
@@ -431,6 +431,25 @@
...
+
+        <listitem><para>Setup this encrypted block device in the initramfs, similarly to
+        <citerefentry><refentrytitle>systemd.mount</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+        units marked with <option>x-initrd.mount</option>.</para>
+
+        <para>Although it's not necessary to mark the mount entry for the root file system with
+        <option>x-initrd.mount</option>, <option>x-initrd.attach</option> is still recommended with
+        the encrypted block device containing the root file system as otherwise systemd will
+        attempt to detach the device during the regular system shutdown while it's still in
+        used. With this option the device will still be detached but lately after the root file
```
Comment (keszybz) [inline man/crypttab.xml]: in used → in use
but lately → but later
after the root file system will be unnounted . → after the root file system is unmounted.

File: man/crypttab.xml [hunk 6]
```diff
@@ -431,6 +431,25 @@
...
+        system will be unmounted .</para>
+
+        <para>All other encrypted block devices that contain file system mounted in the initramfs
```
Comment (keszybz) [inline man/crypttab.xml]: file system → file systems

File: man/crypttab.xml [hunk 7]
```diff
@@ -431,6 +431,25 @@
...
       </varlistentry>
 
+      <varlistentry>
+        <term><option>x-initrd.attach</option></term>
+
+        <listitem><para>Setup this encrypted block device in the initramfs, similarly to
+        <citerefentry><refentrytitle>systemd.mount</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+        units marked with <option>x-initrd.mount</option>.</para>
+
+        <para>Although it's not necessary to mark the mount entry for the root file system with
```
Comment (keszybz) [inline man/crypttab.xml:441]: I can't say I like the general recommendation of this paragraph. I agree with it is OK to add this mechanism so it can be used in the corner cases that we can't figure out, but recommending to people to use this pretty much always seems like a step in the wrong direction. Maybe weaken this by saying "This option may be added to any block device that contains a file system which should not be unmounted before switching back to the initramfs during shutdown."

Comment (fbuihuu) [reply man/crypttab.xml:441]: Well unless we start trying to inherit x-initrd.mount in fstab to crypttab somehow we won't be able to figure out which encrypted devices need the option I think. And I don't think we want to do that, do we ?

So it might better at the end to recommend marking all relevant devices rather than documenting that some of them get the option automatically but not the others.

Comment (keszybz) [reply man/crypttab.xml:441]: > unless we start trying to inherit x-initrd.mount in fstab to crypttab somehow we won't be able to figure out which encrypted devices need the option I think. And I don't think we want to do that, do we ?

I'd say that is exactly what we want to do. It might not be easy to implement, but if possible, why not?

Comment (fbuihuu) [reply man/crypttab.xml:441]: That reminds me a old but similar discussion where inheriting options from crypttab to fstab wasn't a goo idea: https://github.com/systemd/systemd/pull/2979

Comment (keszybz) [reply man/crypttab.xml:441]: #2979 was in the opposite direction: propagating settings from low-level units to higher level ones. The upstart model, if you so will. The still support the conclusion to turn that down.

Comment (fbuihuu) [reply man/crypttab.xml:441]: right.

File: units/system-systemd\x2dcryptsetup.slice [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#  SPDX-License-Identifier: LGPL-2.1+
+#
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+[Unit]
```
Comment (arianvp) [inline units/system-systemd\x2dcryptsetup.slice:10]: Drive-by question. I'm a bit confused about this file. As far as I see it is unused. What is the purpose of this slice unit?

Comment (keszybz) [reply units/system-systemd\x2dcryptsetup.slice:10]: @fbuihuu ?

Comment (poettering) [reply units/system-systemd\x2dcryptsetup.slice:10]: all instance units of the same template implicitly get their own slice assigned covering all instances. This unit file just attaches a nice description to that slice unit, which is a unit that is synthesized anyway, but otherwise wouldn't have a nice description

It might make sense we carefully makes sure we do the same for all such implicit slices synthesized for multi-instantiated units.

Comment (keszybz) [reply units/system-systemd\x2dcryptsetup.slice:10]: It's used automatically by `systemd-cryptsetup@.service`. But the commit doesn't say why this is needed, and even though the unit file refers to systemd.special(7), the unit is not mentioned there. I'm not sure if this was unit was left over from an earlier version of the patchset, or if actually serves a purpose.

Comment (keszybz) [reply units/system-systemd\x2dcryptsetup.slice:10]: So... is the Description line the only purpose? I'd just drop it then, it doesn't seem worth it, and we don't do it for other units.

Comment (fbuihuu) [reply units/system-systemd\x2dcryptsetup.slice:10]: According to the commit message: "Note that the slice containing the cryptsetup services also needs to drop the
usual shutdown dependencies as it's required by the cryptsetup services."

So the main purpose is to drop the usual shutdown deps I would say. Maybe PID1 got the ability to do that automatically now ?

Comment (keszybz) [reply units/system-systemd\x2dcryptsetup.slice:10]: Oh, so the slice indeed does have `DefaultDependencies=no`.
So it seems all good, except that we should document the slice in systemd.special(7) or remove the Documentation link.

Comment (keszybz) [reply units/system-systemd\x2dcryptsetup.slice:10]: https://github.com/systemd/systemd/pull/19260
